File: src/algorithms/trees/traversal/diagonal-traversal/sources/TreeDiagonalTraversal_test.cpp
#include "TreeDiagonalTraversal.cpp"
#include <cassert>
#include <vector>

BSTNode* makeNode(int value, BSTNode* left = nullptr, BSTNode* right = nullptr) {
    BSTNode* node = new BSTNode(value);
    node->left = left;
    node->right = right;
    return node;
}

int main() {
    TreeDiagonalTraversal sol;

    // balanced 7-node BST
    BSTNode* root1 = makeNode(4,
        makeNode(2, makeNode(1), makeNode(3)),
        makeNode(6, makeNode(5), makeNode(7)));
    std::vector<std::vector<int>> expected1 = {{4, 6, 7}, {2, 5, 3}, {1}};
    assert(sol.treeDiagonalTraversal(root1) == expected1);

    // null root
    assert(sol.treeDiagonalTraversal(nullptr).empty());

    // single node
    std::vector<std::vector<int>> expected3 = {{42}};
    assert(sol.treeDiagonalTraversal(makeNode(42)) == expected3);

    // right-skewed tree
    BSTNode* rightSkewed = makeNode(1, nullptr, makeNode(2, nullptr, makeNode(3)));
    std::vector<std::vector<int>> expected4 = {{1, 2, 3}};
    assert(sol.treeDiagonalTraversal(rightSkewed) == expected4);

    // left-skewed tree
    BSTNode* leftSkewed = makeNode(3, makeNode(2, makeNode(1)));
    std::vector<std::vector<int>> expected5 = {{3}, {2}, {1}};
    assert(sol.treeDiagonalTraversal(leftSkewed) == expected5);

    return 0;
}

File: src/algorithms/trees/traversal/diagonal-traversal/__tests__/TreeDiagonalTraversal_test.java
import java.util.List;

public class TreeDiagonalTraversal_test {
    static BSTNode makeNode(int value, BSTNode left, BSTNode right) {
        BSTNode node = new BSTNode(value);
        node.left = left;
        node.right = right;
        return node;
    }

    public static void main(String[] args) {
        TreeDiagonalTraversal sol = new TreeDiagonalTraversal();

        // balanced 7-node BST
        BSTNode root1 = makeNode(4,
            makeNode(2, makeNode(1, null, null), makeNode(3, null, null)),
            makeNode(6, makeNode(5, null, null), makeNode(7, null, null)));
        assert sol.treeDiagonalTraversal(root1).equals(List.of(List.of(4, 6, 7), List.of(2, 5, 3), List.of(1))) : "Test 1 failed";

        // null root
        assert sol.treeDiagonalTraversal(null).isEmpty() : "Test 2 failed";

        // single node
        assert sol.treeDiagonalTraversal(makeNode(42, null, null)).equals(List.of(List.of(42))) : "Test 3 failed";

        // right-skewed tree
        BSTNode rightSkewed = makeNode(1, null, makeNode(2, null, makeNode(3, null, null)));
        assert sol.treeDiagonalTraversal(rightSkewed).equals(List.of(List.of(1, 2, 3))) : "Test 4 failed";

        // left-skewed tree
        BSTNode leftSkewed = makeNode(3, makeNode(2, makeNode(1, null, null), null), null);
        assert sol.treeDiagonalTraversal(leftSkewed).equals(List.of(List.of(3), List.of(2), List.of(1))) : "Test 5 failed";

        System.out.println("All tests passed!");
    }
}
